src/main/kotlin/org/jetbrains/research/libsl/visitors/TopLevelDeclarationsResolver.kt

```diff
@@ -33,13 +53,49 @@ class TopLevelDeclarationsResolver(
     }
 
     override fun visitVariableDecl(ctx: LibSLParser.VariableDeclContext) {
+        val keyword = VariableKeyword.fromString(ctx.keyword.text)
         val variableName = ctx.nameWithType().name.text.extractIdentifier()
         val typeRef = processTypeIdentifier(ctx.nameWithType().type)
 
         val expressionVisitor = ExpressionVisitor(context)
         val initialValue = ctx.assignmentRight()?.let { expressionVisitor.visitAssignmentRight(it) }
 
-        val variable = VariableWithInitialValue(variableName, typeRef, initialValue)
+        val annotationUsages = getAnnotationUsages(ctx.annotationUsage())
+        val variable = VariableWithInitialValue(
+            keyword,
+            variableName,
+            typeRef,
+            annotationUsages,
+            initialValue
+        )
         globalContext.storeVariable(variable)
     }
+
+    override fun visitActionDecl(ctx: LibSLParser.ActionDeclContext) {
+        val actionName = ctx.actionName.text.extractIdentifier()
+        val actionParams = mutableListOf<DeclaredActionParameter>()
+
+        ctx.actionDeclParamList().actionParameter().forEach { param ->
+            val actionParam = DeclaredActionParameter(
+                param.name.text.extractIdentifier(),
+                processTypeIdentifier(param.type),
+                getAnnotationUsages(param.annotationUsage())
+            )
+
+            actionParams.add(actionParam)
+        }
+
+        val returnType = ctx.actionType?.let { processTypeIdentifier(it) }
+
+        if (returnType != null) {
+            val resultVariable = ResultVariable(returnType)
```

Comment:
Why do you need result variable here? And why do you store it in global context?

Reply:
Removed